lib/pbench/common/utils.py

```diff
@@ -49,6 +49,11 @@ def md5sum(filename: Union[Path, str]) -> Md5Result:
     flags=re.IGNORECASE,
 )
 
+_satellite_allowed = re.compile(
+    r"[A-Za-z0-9]+::[A-Z0-9](?:[A-Z0-9\-]{0,61}[A-Z0-9])?(?:\.[A-Z0-9](?:[A-Z0-9\-]{0,61}[A-Z0-9])?)*",
+    flags=re.IGNORECASE,
+)
+
```

Comment:
While I appreciate that you duplicated the expression to simplify review and discussion, if we finalize this we should simply update a single regex rather than having two which are always used together in an `or` as below. 😀 

Also, note (though this isn't relevant to the immediate concern) that this doesn't support IPv6 addresses, and we really should. (Eventually, that would be simplified by having a single regex to maintain rather than two near duplicates.)

We could also consider, instead, simply phasing out the `controller` header and validation thereof from `PUT`. We still use it as an intermediate directory in the filesystem, but `PUT` can get the agent's controller value from `metadata.log`. And, in any case, I see no reason that the *server* ought to be validating it as a hostname. That's an *agent* semantic constraint, which is fine, but I see no value in making it a *server* constraint ... it's just a metadata value. (In fact, even historically, if we were going to validate it'd have been more meaningful to validate as a filename string rather than as a hostname, since that's how we use it!)

Reply:
Historically, we use the satellite prefix to ensure that somebody with a host name inside the fire wall and somebody with the same host name outside the firewall did not get their tar balls put into the same directory.

At least, that was the reason I remember.

It might be best to have PUT shim strip the satellite name off of the controller name, since the original agent did not ask for that, and add it as a piece of metadata via a PUT API header.

Coupled with Dave's suggestion of phasing out the controller header and validation thereof from `PUT`, we'd have the added satellite metadata added without modifying the tar ball contents on the `PUT`.

Reply:
And of course another possibility is for the put-shim server to drop the prefix (which it defines and owns, through the satellite sync mechanism) before specifying the `controller` header on `PUT`. I do, however, like the idea of having the 1.0 server understand it was propagated through the satellite.

I've thought of adding a metadata mechanism to `PUT` (e.g., `?metadata=key:value,key:value`), and we could define something like `global.pbench.satellite:RIYA`, which could be searched. Or given that we need "API keys" (which we haven't fully defined much less implemented), we could grant the put-shim server a special role allowing it to set a "privileged" key, for example `server.satellite:RIYA`. This would allow adding a `--metadata` option to `pbench-results-move` (and friends), which would be passed through; and put-shim could leverage that. Then the satellite origin would be metadata, as it should be, rather than mucking up the "controller" namespace assumptions.

Reply:
Things that make you go, hmmm...

Reply:
> Things that make you go, hmmm...

The problem with GitHub review comments is that one can only guess the referents. My second comment and your first show 1 minute difference in timestamps, I didn't see yours before I wrote mine, and I can only guess whether you saw my second before writing your second... much less guess what specifically made you go "hmmm" although there seem on the surface to be many things in the PR and each comment which might have had that effect. 😆

Reply:
I believe I concur with each of the preceding.  😁

To extend Dave's initial comment, I believe that extending the previous expression rather than adding a second would be better, to wit:
```
_allowed = re.compile(
    r"(?:[A-Za-z0-9]+::)?[A-Z0-9](?:[A-Z0-9\-]{0,61}[A-Z0-9])?(?:\.[A-Z0-9](?:[A-Z0-9\-]{0,61}[A-Z0-9])?)*",
    flags=re.IGNORECASE,
)
```
That is, we allow it if it meets the previous constraints with the added allowance that it may be prefixed by a string of letters and numbers ending in a double-colon.

Reply:
Ultimately, I think we should do here what we are going to do when the DMZ solution is in place. Will it be the case that whether a dataset originated inside or outside the firewall, the server will not be able to tell where it came from? If so, there is no point in making the distinction now either - no? It'll just be a label on datasets that were collected in this (the satellite -> put-shim) way for the limited period of time that the mechanism will be in operation - not much point in it, AFAICT.

Historically, I think it was meant as a way to disambiguate datasets from controllers with the same name, but in practice, it didn't do a very good job of that either: the main problem was that "localhost" was the most common controller name both inside and outside the firewall, so instead of one huge pile, we got two smaller but still huge piles of datasets...

Reply:
A good part of the value is that the dataset is passing through multiple servers to get to its final resting place (RIP): there's some diagnostic value in providing a visible trace.

I like the idea of recording metadata because it came from a satellite; and in fact we ought to be recording as well that it passed through the put-shim server. Once I get `PUT` able to add metadata, it'll be easy for put-shim to do both.

If our final solution (as seems likely) is to allow the server to reach out to an external S3 service with a `GET` from a designated object ID, we might want to consider tracking *that* with metadata, as well.